File: frontend/src/lib/bus-settings.ts
export const DEFAULT_BUS_PARTICIPANT_LIMITS: Record<string, number> = {
  can: 64, can_fd: 64, can_xl: 64, lin: 64, automotive_ethernet: 256, flexray: 64,
};
export const BUS_SETTING_LABELS: Record<string, string> = {
  can: "CAN", can_fd: "CAN-FD", can_xl: "CAN-XL", lin: "LIN", automotive_ethernet: "Ethernet", flexray: "FlexRay",
};
export function busSettingKey(value: string): string {
  const key = value.toLowerCase().replace(/^detected:/, "").replace(/[^a-z0-9]/g, "");
  return ({canfd: "can_fd", canxl: "can_xl", ethernet: "automotive_ethernet", automotiveethernet: "automotive_ethernet", someip: "automotive_ethernet"} as Record<string, string>)[key] ?? key;
}
export function normalizeBusLimits(value: unknown): Record<string, number> {
  const result = {...DEFAULT_BUS_PARTICIPANT_LIMITS};
  if (value && typeof value === "object" && !Array.isArray(value)) {
    for (const [key, limit] of Object.entries(value)) {
      if (busSettingKey(key) in result && Number.isInteger(limit) && (limit === 0 || limit >= 2) && limit <= 100000) result[busSettingKey(key)] = limit;
    }
  }
  return result;
}
export function busBranchCapacity(limits: Record<string, number>, technology: string): number {
  const limit = limits[busSettingKey(technology)] ?? 0;
  return limit ? limit - 1 : 100000;
}

File: backend/engineering/bus_settings.py
"""Project planning limits. These are configurable simulation inputs, not protocol claims."""
from __future__ import annotations

import json
import re

DEFAULT_BUS_PARTICIPANT_LIMITS = {
    'can': 64, 'can_fd': 64, 'can_xl': 64, 'lin': 64,
    'automotive_ethernet': 256, 'flexray': 64,
}


def bus_key(value: str) -> str:
    key = re.sub(r'[^a-z0-9]', '', str(value).lower().removeprefix('detected:'))
    return {'canfd': 'can_fd', 'canxl': 'can_xl', 'ethernet': 'automotive_ethernet',
            'automotiveethernet': 'automotive_ethernet', 'someip': 'automotive_ethernet'}.get(key, key)


def normalize_bus_limits(value=None) -> dict[str, int]:
    result = dict(DEFAULT_BUS_PARTICIPANT_LIMITS)
    if value is None:
        return result
    if not isinstance(value, dict):
        raise ValueError('Bus-Teilnehmergrenzen müssen ein Objekt sein.')
    for technology, limit in value.items():
        key = bus_key(technology)
        if key not in result:
            raise ValueError(f'Unbekannter Bustyp: {technology}')
        if isinstance(limit, bool) or not isinstance(limit, int) or limit < 0 or limit == 1 or limit > 100000:
            raise ValueError('Teilnehmergrenzen müssen 0 (unbegrenzt) oder ganze Zahlen von 2 bis 100000 sein.')
        result[key] = limit
    return result


def limits_from_prompt(prompt: str) -> dict[str, int]:
    match = re.search(r'^- Bus-Teilnehmergrenzen:\s*(\{[^\r\n]*\})\s*$', prompt, re.M)
    return normalize_bus_limits(json.loads(match.group(1)) if match else None)


def branch_capacity(limits: dict, technology: str) -> int:
    """Reserve one participant for the gateway/controller on a shared segment."""
    total = limits.get(bus_key(technology), 0)
    return total - 1 if total else 100000


def with_project_limits(prompt: str, context: dict) -> str:
    """Freeze planning settings into an agent request unless already specified."""
    if re.search(r"^- Bus-Teilnehmergrenzen:", prompt, re.M):
        limits_from_prompt(prompt)
        return prompt
    settings = context.get("engineering_wizard_settings") or {}
    limits = normalize_bus_limits(settings.get("bus_participant_limits"))
    return prompt.rstrip() + "\n- Bus-Teilnehmergrenzen: " + json.dumps(limits, sort_keys=True) + "\n"
